### PR TITLE
Rework of guard stats, including additional HP and different heart co…

### LINKED ISSUE #2975: Guards need improvement
### Minecolonies version

Version: 1.12.2-0.10.1-RELEASE

### Expected behavior

Guards to be able to handle themselves even in a single encounter. Are they supposed to be able to kill enemies or are they just for looks?

### Actual behavior

Guards, even when you give them armor, are always dying.  

### Steps to reproduce the problem

Equip your army of guards with weapons and armor and let night come.

Watch the slaughter of your guards.

I know this is possibly a touchy subject but these guys may need some sort of boost or something... particularly when you upgrade their towers maybe?  It just seems that they are way too squishy and do nothing but either eat up resources or spam the chat when you dont give them anything.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/64730011-guards-need-improvement?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by someaddons**
Closes #2975

# Changes proposed in this pull request:
-Added some level scaling for HP/Damage, and adjusted current scaling(attackspeed and such)
-Added Configuration multipliers for Guards damage/HP
-Added colored hearts as a way to show more HP, goes from red->golden->green->blue and can show up to 80 hp
-Ranged guard's firing from too far/burning through bows fixed
-Ranged logic cleanup/bugifx/Reducing tick rate to 2 tps from 20.
Review please

### CODE REVIEW

File: src/api/java/com/minecolonies/api/configuration/Configurations.java [hunk 1]
```diff
@@ -201,6 +201,15 @@
         @Config.Comment("Should Guard Rangers benefit from Power/Smite/Bane of Arthropods enchants?")
         public boolean rangerEnchants = true;
 
+        @Config.Comment("Damage multiplier for Ranger Guards: Default:1.0")
+        public float rangerDamageMult = 1.0f;
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/configuration/Configurations.java]: I feel like we should make these doubles

File: src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java [hunk 2]
```diff
@@ -740,6 +743,56 @@ else if (citizenData.getSaturation() < LOW_SATURATION)
...
+        }
+        if (getHealth() > getMaxHealth())
+        {
+            setHealth(getMaxHealth());
+        }
+    }
+
+    /**
+     * Removes a maxhealth modifier by name
+     */
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:780]: Missing the javadoc @param

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 3]
```diff
@@ -300,6 +304,9 @@ public AIState preAttackChecks()
 
         if(target == null || target.isDead || !isWithinPersecutionDistance(target.getPosition()))
         {
+            // Clear pathing when target changes
+            worker.getNavigator().clearPath();
+            worker.getMoveHelper().strafe(0, 0);
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java:308]: Good find

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java [hunk 4]
```diff
@@ -414,6 +423,7 @@ else if (entity instanceof EntityCitizen && colony.isValidAttackingGuard((Entity
 
     /**
      * Check if a position is within regular attack distance.
+     * TODO:Fix so it actually says if sth is within attack Dist.
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java:425]: Future TODO or forgot to remove it?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIRanger.java [hunk 5]
```diff
@@ -102,10 +98,32 @@ public AIState getAttackState()
...
     @Override
     protected int getAttackRange()
     {
-        return (MAX_DISTANCE_FOR_ATTACK / ((buildingGuards.getBuildingLevel() / buildingGuards.getMaxBuildingLevel()) + 1)) + 10;
+        return getRealAttackRange() + 10;
+    }
+
+    /**
+     * Calculates the actual attack range
+     */
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIRanger.java:113]: @returns ?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIRanger.java [hunk 6]
```diff
@@ -102,10 +98,32 @@ public AIState getAttackState()
...
+    private int getRealAttackRange()
+    {
+        int attackDist = BASE_DISTANCE_FOR_RANGED_ATTACK;
+        // + 2 Blockrange per building level for a total of +10 from building level
+        if (buildingGuards != null)
+        {
+            attackDist += buildingGuards.getBuildingLevel() * 2;
+        }
+        // ~ +1 each three levels for a total of +15 from guard level
+        if (worker.getCitizenData() != null)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIRanger.java:123]: I like how it scales.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java (hunk 4 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java:425]: ye future todo, didnt wanna add it to this one

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java:425]: Yeah, it's pretty big already.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIRanger.java (hunk 6 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/EntityAIRanger.java:123]: :) figured sclaing with hut level + guard level would be nice

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java (hunk 3 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java:308]: 👍

**Review by marchermans — DISMISSED** [review]

**Pull request merged**